[Part1] Implemented Thunder method for SVM algorithm on CPU

PetrovKP opened the pull request:
1. Implemented method Thunder SVM on CPU. Reference: Zeyi Wen, Jiashuai Shi, Bingsheng He; ThunderSVM: A Fast SVM Library on GPUs and CPUs, Journal of Machine Learning Research, 19, 1-5 (2018). [Link](https://github.com/Xtra-Computing/thundersvm/blob/master/thundersvm-full.pdf)
2. Refactoring Boser method for use SMO Local kernels and save result for all methods. 
3. Added Safe dispatcher for OneAPI kernel wrappers. Added dispatcher for thunder method.

SmirnovEgorRu commented:
Amazing progress, I appreciate this.

PetrovKP commented:
/intelci: run

PetrovKP commented:
/intelci: run

PetrovKP commented:
/intelci: run

PetrovKP commented:
/intelci: run

PetrovKP commented:
/intelci: run

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented on algorithms/kernel/svm/svm_train_common_avx512_impl.i:
```diff
@@ -68,36 +67,36 @@ void SVMTrainTask<float, daal::algorithms::svm::interface2::Parameter, avx512>::
-
-        // vector multiplication y_grad[jcur:jcur+16] = _y[jcur:jcur+16] * _grad[jcur:jcur+16]
-        __m512 valYGrad = _mm512_mul_ps(_mm512_loadu_ps(&_y[j_cur]), _mm512_loadu_ps(&_grad[j_cur]));
+        __m128i vec_I = _mm_load_si128((__m128i *)(&I[j_cur]));
```
Thanks

PetrovKP replied:
Done

PetrovKP commented on algorithms/kernel/svm/svm_train_common_avx512_impl.i:
```diff
@@ -109,101 +108,99 @@ void SVMTrainTask<float, daal::algorithms::svm::interface2::Parameter, avx512>::
     for (size_t k = 0; k < 16; ++k)
     {
-        if (minus_GMin_val[k] > GMin)
+        if (minus_GMax_val[k] > GMax)
```
Here reduction for 16 elements. I do not remember. Are there instructions for reduction inside vector registers?

SmirnovEgorRu replied:
Do you mean ```_mm512_reduce_max_ps```?

PetrovKP replied:
I'll try, thanks!

PetrovKP replied:
It looks cool, but I did not find instructions that would help me find the argmax. So far I leave it like this -  this code should not bring in much work.

SmirnovEgorRu replied:
to find index, you need:
auto max = _mm512_reduce_max_ps(a);
auto max_reg = _mm512_set1_ps(max);
auto mask = _mm512_cmpeq_ps(a, max_reg);
int index = _mm_tzcnt_32(mask);

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented:
@Vika-F , @agorshk Please, look changes 1,2:

1. Implemented method Thunder SVM on CPU. Reference: Zeyi Wen, Jiashuai Shi, Bingsheng He; ThunderSVM: A Fast SVM Library on GPUs and CPUs, Journal of Machine Learning Research, 19, 1-5 (2018). [Link](https://github.com/Xtra-Computing/thundersvm/blob/master/thundersvm-full.pdf)
2. Refactoring Boser method for use SMO Local kernels and save result for all methods.

@masdevas Please, look change 3:

3. Added Intel dispatcher for OneAPI kernel wrappers. Added dispatcher for thunder method.

masdevas commented on algorithms/kernel/kernel.h:
```diff
@@ -175,7 +175,7 @@
         : BaseClassName(daalEnv), _cntr(NULL)                                                                                                       \
     {                                                                                                                                               \
-        switch (daalEnv->cpuid)                                                                                                                     \
+        GetCpuid switch (daalEnv->cpuid)                                                                                                            \
```
Please change "daalEnv->cpuid" to "cpuid"

PetrovKP replied:
Thanks!

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented on algorithms/kernel/svm/svm_train_common_avx512_impl.i:
```diff
@@ -109,101 +108,99 @@ void SVMTrainTask<float, daal::algorithms::svm::interface2::Parameter, avx512>::
+    _mm512_storeu_ps(minus_GMax_val.get(), valMinusGMax);
     _mm512_storeu_si512(Bj_val_overall.get(), Bj_vec);
-    _mm512_storeu_ps(minusGMin2_overall.get(), valMinusGMin2);
+    _mm512_storeu_ps(minusGMax2_overall.get(), valMinusGMax2);
```
Done

PetrovKP commented:
@ShvetsKS Please, look this PR.

ShvetsKS commented on algorithms/kernel/svm/svm_train_boser_cache.i:
```diff
@@ -44,27 +46,16 @@ using namespace daal::internal;
     virtual ~SVMCacheIface() {}
 
-    virtual size_t getDataRowIndex(size_t rowIndex) const = 0;
+    // virtual size_t getDataRowIndex(size_t rowIndex) const = 0;
```
don't forget to delete it

ShvetsKS commented on algorithms/kernel/svm/svm_train_boser_impl.i:
```diff
@@ -514,17 +166,19 @@ void SVMTrainTask<algorithmFPType, ParameterType, cpu>::WSSjLocal(const size_t j
     algorithmFPType fpMax = MaxVal<algorithmFPType>::get();
-    algorithmFPType GMin  = fpMax; // some big positive number
-    algorithmFPType GMin2 = fpMax;
+    algorithmFPType GMax  = -fpMax; // some big positive number
```
comment is not aligned with sign of variable.

PetrovKP replied:
Fixed

ShvetsKS commented on algorithms/kernel/svm/svm_train_common_avx512_impl.i:
```diff
@@ -68,36 +67,36 @@ void SVMTrainTask<float, daal::algorithms::svm::interface2::Parameter, avx512>::
-        // find max(y[j]*grad[j]) instead - here's where we find vector of max's.
+        __m512i Bj_vec_cur = _mm512_set1_epi32((int)(j_cur));
+        // vector grad[jcur:jcur+16] = _grad[jcur:jcur+16]
+        __m512 valGrad = _mm512_loadu_ps(&grad[j_cur]);
```
seems it's better to use `_mm512_load_ps` where it is possible (for internally allocated memory we can guaranty alignment)

ShvetsKS commented on algorithms/kernel/svm/svm_train_common_avx512_impl.i:
```diff
@@ -109,101 +108,95 @@ void SVMTrainTask<float, daal::algorithms::svm::interface2::Parameter, avx512>::
-    TNArray<float, 16, avx512> minus_GMin_val(16);
-    TNArray<int, 16, avx512> Bj_val_overall(16);
-    TNArray<float, 16, avx512> minusGMin2_overall(16);
+    float * GMaxVal      = reinterpret_cast<float *>(&valGMax);
```
minor comment to follow codding style:
`float * const GMaxVal `
similar for bellow

ShvetsKS commented on algorithms/kernel/svm/svm_train_common_avx512_impl.i:
```diff
@@ -237,28 +230,28 @@ void SVMTrainTask<double, daal::algorithms::svm::interface2::Parameter, avx512>:
         // a_tmp = two * KiBlock[j - jStart] - kernelDiag[j]
         __m512d KiBlock_vec    = _mm512_loadu_pd(KiBlock + j_cur - jStart);
-        __m512d kernelDiag_vec = _mm512_loadu_pd(&_kernelDiag[j_cur]);
+        __m512d kernelDiag_vec = _mm512_loadu_pd(&kernelDiag[j_cur]);
```
seems for `kernelDiag` we can also guarantee alignment

ShvetsKS commented on algorithms/kernel/svm/svm_train_common_impl.i:
```diff
@@ -0,0 +1,198 @@
+        {
+            a = tau;
+        }
+        algorithmFPType dt      = b / a;
```
const

ShvetsKS replied:
similar for whole block

ShvetsKS commented on algorithms/kernel/svm/svm_train_thunder_cache.h:
```diff
@@ -0,0 +1,217 @@
+            _isComputeSubKernel = true;
+        }
+        const algorithmFPType * const dataIn = _cache.get() + _nSelectRows * _lineSize;
+        algorithmFPType * dataOut            = _cache.get();
```
general comment, follow `algorithmFPType * const dataOut`

ShvetsKS commented on algorithms/kernel/svm/svm_train_common.h:
```diff
@@ -0,0 +1,171 @@
+    {
+        const size_t p                  = xTable->getNumberOfColumns();
+        const size_t nRows              = xTable->getNumberOfRows();
+        CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(xTable.get()));
```
should this pointer be checked on 'nullptr' ?

PetrovKP replied:
Fixed

ShvetsKS commented on algorithms/kernel/svm/svm_train_common_impl.i:
```diff
@@ -0,0 +1,198 @@
+services::Status SubDataTaskCSR<algorithmFPType, cpu>::copyDataByIndices(const uint32_t * wsIndices, const NumericTablePtr & xTable)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(cache.copyDataByIndices.CSR);
+    CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(xTable.get()));
```
should this pointer be checked on nullptr?

PetrovKP replied:
Applied

ShvetsKS commented on algorithms/kernel/svm/svm_train_common_impl.i:
```diff
@@ -0,0 +1,198 @@
+        const algorithmFPType * values   = mtX.values();
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t j = 0; j < nNonZeroValuesInRow; j++, dataIndex++)
```
why don't use `daal_memcpy` here, and then displace `dataIndex` (`dataIndex+=nNonZeroValuesInRow`)?

ShvetsKS commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,338 @@
+    TArray<char, cpu> I(nWS);
+    DAAL_CHECK_MALLOC(I.get());
+
+    if (cacheSize >= nVectors * nVectors * sizeof(algorithmFPType))
```
`nVectors * nVectors`  we have to check such computations on overflow according to SDL requirements.

ShvetsKS commented on algorithms/kernel/svm/svm_train_thunder_kernel.h:
```diff
@@ -0,0 +1,90 @@
+    // The maximum block size for blocked SMO solver.
+    // Need of (maxBlockSize*5 + maxBlockSize*maxBlockSize)*sizeof(algorithmFPType) internal memory.
+    // It should get into the cache L2 (~1MB).
+    static const size_t maxBlockSize = 1024;
```
why don't we want to calculate L2 sizes via utils functions to take into consideration different number columns?

PetrovKP replied:
We are not dependent on anything. Only from `maxBlockSize `, because we copy in local buffers.

ShvetsKS replied:
So possibly it can be a user defined parameter. Make sense to consider it in future.

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

averbukh commented on algorithms/kernel/kernel_function/kernel_function_linear_dense_default_impl.i:
```diff
@@ -132,39 +132,10 @@ services::Status KernelImplLinear<defaultDense, algorithmFPType, cpu>::computeIn
-                                                                             (DAAL_INT *)&nFeatures, &beta, r, (DAAL_INT *)&nVectors2);
-        if (!retStat) return retStat;
-    }
+    services::Status retStat =
```
const ?

averbukh commented on algorithms/kernel/svm/inner/svm_train_batch_container_v1.h:
```diff
@@ -67,10 +67,22 @@ services::Status BatchContainer<algorithmFPType, method, cpu>::compute()
     algorithms::Model * r = static_cast<daal::algorithms::Model *>(result->get(classifier::training::model).get());
 
-    svm::interface1::Parameter * par = static_cast<svm::interface1::Parameter *>(_par);
+    svm::interface1::Parameter * par1 = static_cast<svm::interface1::Parameter *>(_par);
```
const * const ?

averbukh commented on algorithms/kernel/svm/svm_train_common.h:
```diff
@@ -0,0 +1,172 @@
+    {
+        const size_t p                  = xTable->getNumberOfColumns();
+        const size_t nRows              = xTable->getNumberOfRows();
+        CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(xTable.get()));
```
CSRNumericTableIface * const

averbukh commented on algorithms/kernel/svm/svm_train_common.h:
```diff
@@ -0,0 +1,172 @@
+        CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(xTable.get()));
+        if (!csrIface) return;
+        ReadRowsCSR<algorithmFPType, cpu> mtX(csrIface, 0, nRows);
+        const size_t * rowOffsets = mtX.rows();
```
const size_t * const

averbukh commented on algorithms/kernel/svm/svm_train_common_impl.i:
```diff
@@ -0,0 +1,209 @@
+    algorithmFPType GMin = (MaxVal<algorithmFPType>::get()); // some big negative number
+
+    /* Find i index of the working set (Bi) */
+    for (size_t i = 0; i < nActiveVectors; i++)
```
++i instead of i++

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+
+        const algorithmFPType zero(0.0);
+        size_t nSV = 0;
+        for (size_t i = 0; i < _nVectors; i++)
```
++i

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+
+        WriteOnlyRows<int, cpu> mtSvIndices(*svIndicesTable, 0, nSV);
+        DAAL_CHECK_BLOCK_STATUS(mtSvIndices);
+        int * svIndices = mtSvIndices.get();
```
int * const ?

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+        int * svIndices = mtSvIndices.get();
+
+        const algorithmFPType zero(0.0);
+        for (size_t i = 0, iSV = 0; i < _nVectors; i++)
```
++i

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+            {
+                DAAL_ASSERT(_cache->getDataRowIndex(i) <= services::internal::MaxVal<int>::get())
+                svIndices[iSV] = (int)_cache->getDataRowIndex(i);
+                iSV++;
```
++iSV;

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+
+        WriteOnlyRows<algorithmFPType, cpu> mtSv(*svTable, 0, nSV);
+        DAAL_CHECK_BLOCK_STATUS(mtSv);
+        algorithmFPType * sv = mtSv.get();
```
const ?

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+
+        TArray<size_t, cpu> aSvRowOffsets(nSV + 1);
+        DAAL_CHECK_MALLOC(aSvRowOffsets.get());
+        size_t * svRowOffsetsBuffer = aSvRowOffsets.get();
```
const ?

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+        DAAL_CHECK_MALLOC(aSvRowOffsets.get());
+        size_t * svRowOffsetsBuffer = aSvRowOffsets.get();
+
+        CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(&xTable));
```
check dynamic_cast result?

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+                mtX.set(csrIface, rowIndex, 1);
+                DAAL_CHECK_BLOCK_STATUS(mtX);
+                svRowOffsetsBuffer[iSV + 1] = svRowOffsetsBuffer[iSV] + (mtX.rows()[1] - mtX.rows()[0]);
+                iSV++;
```
prefix increment

averbukh commented on algorithms/kernel/svm/svm_train_result.h:
```diff
@@ -0,0 +1,335 @@
+                sv[svOffset]           = xi[j];
+                svColIndices[svOffset] = xiColIndices[j];
+            }
+            iSV++;
```
prefix increment

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_cache.h:
```diff
@@ -0,0 +1,217 @@
+ * Common interface for cache that stores kernel function values
+ */
+template <typename algorithmFPType, CpuType cpu>
+class SVMCacheIface<thunder, algorithmFPType, cpu> : public SVMCacheCommonIface<algorithmFPType, cpu>
```
Class name confuses a little bit, because class contains data fields while have Iface in the name.

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+    // gradi = -yi; ai = 0
+    TArray<algorithmFPType, cpu> gradTArray(nVectors);
+    DAAL_CHECK_MALLOC(gradTArray.get());
+    algorithmFPType * grad = gradTArray.get();
```
const pointer?

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    size_t iter = 0;
+    for (; iter < maxIterations; iter++)
```
++iter

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+        cachePtr                      = SVMCache<thunder, noCache, algorithmFPType, cpu>::create(defaultCacheSize, nVectors, xTable, kernel, status);
+    }
+
+    size_t iter = 0;
```
Makes sense to combine with next line ?

PetrovKP replied:
I would like to leave it because of debugging the number of iterations. In the future, I would like to return as a result.

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+
+    const size_t innerMaxIterations(nWS * cInnerIterations);
+
+    algorithmFPType * alphaLocal    = buffer + nWS * MemSmoId::alphaBuffID;
+    algorithmFPType * yLocal        = buffer + nWS * MemSmoId::yBuffID;
+    algorithmFPType * gradLocal     = buffer + nWS * MemSmoId::gradBuffID;
+    algorithmFPType * kdLocal       = buffer + nWS * MemSmoId::kdBuffID;
+    algorithmFPType * oldAlphaLocal = buffer + nWS * MemSmoId::oldAlphaBuffID;
+    algorithmFPType * kernelLocal   = buffer + nWS * MemSmoId::latest;
```
Const pointers ?

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+            const size_t startRow = iBlock * blockSize;
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = startRow; i < startRow + blockSize; i++)
```
++i

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+                Ii |= HelperTrainSVM<algorithmFPType, cpu>::isUpper(yLocal[i], alphaLocal[i], C) ? up : free;
+                Ii |= HelperTrainSVM<algorithmFPType, cpu>::isLower(yLocal[i], alphaLocal[i], C) ? low : free;
+                I[i] = Ii;
+                for (size_t j = 0; j < nWS; j++)
```
++j

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,379 @@
+    DAAL_CHECK_MALLOC(gradTArray.get());
+    algorithmFPType * grad = gradTArray.get();
+
+    for (size_t i = 0; i < nVectors; i++)
```
Add comment if so, because every first will try to fix

PetrovKP replied:
I want to add support for standard labels in DAAL (0,1). In 3 parts I will add this and all initialization to the parallel regions.

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+    DAAL_INT lda = m_;
+    DAAL_INT incx(1);
+    algorithmFPType beta(1.0);
+    DAAL_INT incy(1);
```
incY. At least to avoid confusion ('incy' is valid word).

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_impl.i:
```diff
@@ -0,0 +1,339 @@
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(updateGrad);
+    char trans  = 'N';
+    DAAL_INT m_ = nVectors;
```
Do you need '_' in the name?

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_workset.h:
```diff
@@ -0,0 +1,201 @@
+                {
+                    _wsIndices[_nSelected] = i;
+                    _indicator[i]          = true;
+                    _nSelected++;
```
use prefix increment

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_workset.h:
```diff
@@ -0,0 +1,201 @@
+    const IndexType * getIndices() const { return _wsIndices.get(); }
+
+protected:
+    size_t maxpow2(size_t n)
```
Consider better function name to avoid confusion

PetrovKP replied:
Applied

averbukh commented on algorithms/kernel/svm/svm_train_thunder_workset.h:
```diff
@@ -0,0 +1,201 @@
+        }
+
+        size_t count = 0;
+        while (n > 1)
```
Looks like you can use intrinsic instead of this loop

PetrovKP replied:
May be. But this function is called once for the entire algorithm call. Its time is minimal - I would not want to embed the instincts in this place.

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu

PetrovKP commented:
/intelci: run
set infra_branch=dev/svm-thunder-cpu
